### PR TITLE
drivers: fpga: Add support for ZynqMP FPGA controller

### PR DESCRIPTION
**Pull request by msierszulski**
As mentioned in #36754, we are extending the FPGA controller subsystem.

This PR adds a zynqmp FPGA driver and support for the `mercury_xu` board which uses it to enable the FPGA controller subsystem. It enables FPGA programming from Cortex R on Zynq US+ MPSoC.

The changes consist of adding an FPGA node to `mercury_xu` dts and an entirely new driver.

### CODE REVIEW

**Review by tgorochowik — DISMISSED** [review]

File: drivers/fpga/fpga_zynqmp.c [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+}
+
+static int zynqmp_fpga_reset(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* Reset PL */
+	PCAP_PROG = PCAP_PROG_RESET_MASK;
+	PCAP_PROG = ~PCAP_PROG_RESET_MASK;
+
```
Comment (kgugala) [inline drivers/fpga/fpga_zynqmp.c:255]: shouldn't we check here if the PL is out of reset? `pcap_status` register provides such info.

**Review by kgugala — CHANGES_REQUESTED** [review]
One small comment, otherwise LGTM

File: drivers/fpga/fpga_zynqmp.c (hunk 1 above)
Comment (msierszulski) [reply drivers/fpga/fpga_zynqmp.c:255]: Yes, we should. I added a check to see if PL is out of reset. Thank you!

File: drivers/fpga/fpga_zynqmp.h [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+#define ZU28_IDCODE       0x47E0093
+#define ZU29_IDCODE       0x47E2093
+#define ZU39_IDCODE       0x47E6093
+#define ZU43_IDCODE       0x47FD093
+#define ZU46_IDCODE       0x47F8093
+#define ZU47_IDCODE       0x47FF093
+#define ZU48_IDCODE       0x47FB093
+#define ZU49_IDCODE       0x47FE093
+
+#define LENGTH(data) (*(data + 1) | *data << 0x8U);
```
Comment (tgorochowik) [inline drivers/fpga/fpga_zynqmp.h:69]: Can we give it a more specific name? What exactly is this calculating and how?

Comment (msierszulski) [reply drivers/fpga/fpga_zynqmp.h]: Of course. I renamed it to `XLNX_BITSTREAM_SECTION_LENGTH`. This concatenates the 2 bytes where the given section length is located in the bitstream header.

**Review by tgorochowik — APPROVED** [review]

File: drivers/fpga/fpga_zynqmp.c (hunk 1 above)
Comment (kgugala) [reply drivers/fpga/fpga_zynqmp.c:255]: thanks

**Review by kgugala — APPROVED** [review]

**Pull request merged**